pyproject.toml

```diff
@@ -1,12 +1,146 @@
+    "Programming Language :: Python :: 3.10",
+    "Programming Language :: Python :: Implementation :: CPython",
+    "Topic :: Scientific/Engineering",
+    "Topic :: Scientific/Engineering :: Physics",
+]
+dependencies = [
+    "click>=8.0.0",  # for console scripts
+    "importlib_resources>=1.4.0; python_version < '3.9'",  # for resources in schema
+    "jsonpatch>=1.15",
+    "jsonschema>=4.15.0",  # for utils
+    "pyyaml>=5.1",  # for parsing CLI equal-delimited options
+    "scipy>=1.5.1",  # requires numpy, which is required by pyhf and tensorflow
+    "tqdm>=4.56.0",  # for readxml
+    "numpy",  # compatible versions controlled through scipy
+]
+
+[project.scripts]
+pyhf = "pyhf.cli:cli"
+
+[project.urls]
+Documentation = "https://pyhf.readthedocs.io/"
+Homepage = "https://github.com/scikit-hep/pyhf"
+"Issue Tracker" = "https://github.com/scikit-hep/pyhf/issues"
+"Release Notes" = "https://pyhf.readthedocs.io/en/stable/release-notes.html"
+"Source Code" = "https://github.com/scikit-hep/pyhf"
+
+[project.optional-dependencies]
+shellcomplete = ["click_completion"]
+tensorflow = [
+    "tensorflow>=2.7.0",  # c.f. PR #1962
+    "tensorflow-probability>=0.11.0",  # c.f. PR #1657
+]
+torch = ["torch>=1.10.0"]  # c.f. PR #1657
+jax = [
+    "jax>=0.4.1",  # c.f. PR #2079
+    "jaxlib>=0.4.1",  # c.f. PR #2079
+]
+xmlio = ["uproot>=4.1.1"]  # c.f. PR #1567
+minuit = ["iminuit>=2.7.0"]  # c.f. PR #1895
+contrib = [
+    "matplotlib>=3.0.0",
+    "requests>=2.22.0",
+]
+backends = ["pyhf[tensorflow,torch,jax,minuit]"]
+all = ["pyhf[backends,xmlio,contrib,shellcomplete]"]
+
+# Developer extras
+test = [
+    "pyhf[all]",
+    "scikit-hep-testdata>=0.4.11",
+    "pytest>=6.0",
+    "coverage[toml]>=6.0.0",
+    "pytest-mock",
+    "requests-mock>=1.9.0",
+    "pytest-benchmark[histogram]",
+    "pytest-console-scripts",
+    "pytest-mpl",
+    "pydocstyle",
+    "papermill~=2.3.4",
+    "scrapbook~=0.5.0",
+    "jupyter",
+    "graphviz",
+    "pytest-socket>=0.2.0",  # c.f. PR #1917
+]
+docs = [
+    "pyhf[xmlio,contrib]",
+    "sphinx>=5.1.1",  # c.f. https://github.com/scikit-hep/pyhf/pull/1926
+    "sphinxcontrib-bibtex~=2.1",
+    "sphinx-click",
+    "sphinx_rtd_theme",
+    "nbsphinx!=0.8.8",  # c.f. https://github.com/spatialaudio/nbsphinx/issues/620
+    "ipywidgets",
+    "sphinx-issues",
+    "sphinx-copybutton>=0.3.2",
+    "sphinx-togglebutton>=0.3.0",
+    "ipython!=8.7.0",  # c.f. https://github.com/scikit-hep/pyhf/pull/2068
+]
+develop = [
+    "pyhf[test,docs]",
+    "tbump>=6.7.0",
+    "pre-commit",
+    "nox",
+    "codemetapy>=2.3.0",
+]
+
+[tool.hatch.version]
+source = "vcs"
+
+[tool.hatch.version.raw-options]
 local_scheme = "no-local-version"
 
+[tool.hatch.build.hooks.vcs]
+version-file = "src/pyhf/_version.py"
+
+[tool.hatch.build.targets.sdist]
+exclude = [
+    "/src/conftest.py"
+]
+# Needed to properly include src/pyhf/schemas
+# c.f. https://github.com/pypa/hatch/pull/299
+only-include = [
```

Comment:
Okay so with the addition of `only-include` the `schemas` dir is getting properly picked up by the sdist.

I'm a bit confused though given the docs https://hatch.pypa.io/latest/config/build/#packages

> The `packages` option is semantically equivalent to `only-include` (which takes precedence) except that the shipped path will be collapsed to only include the final component.
> 
> So for example, if you want to ship a package `foo` that is stored in a directory `src` you would do:
>
> ```toml
> [tool.hatch.build.targets.wheel]
> packages = ["src/foo"]
> ```

what "final component" means. Though there seems to be no difference in behavior between including or not including

```toml
[tool.hatch.build.targets.wheel]
packages = ["src/pyhf"]
```

Reply:
`packages` is like` only-include` combined with `sources`, where `sources` would be `sources = ["src"]` for this example. It means that `packages = ["src/this/that/foo/bar"]`  would behave like 
```toml
only-include = ["src/this/that/foo/bar"]
sources = ["src/this/that/foo"[
```
i.e. the final path component (`bar`) is the name of the package as included in the final wheel (because it's taken relative to `sources`).

Reply:
```console
$ rm -rf dist && python -m build . && python -m tarfile --list dist/pyhf*.tar.gz | sed 's|[^/]*/||' > /tmp/sdist-without.txt && python -m zipfile --list dist/pyhf*.whl | sed 's|[^/]*/||' > /tmp/wheel-without.txt
$ git revert 4a4da3e0
$ rm -rf dist && python -m build . && python -m tarfile --list dist/pyhf*.tar.gz | sed 's|[^/]*/||' > /tmp/sdist-with.txt && python -m zipfile --list dist/pyhf*.whl | sed 's|[^/]*/||' > /tmp/wheel-with.txt
$ diff /tmp/sdist-without.txt /tmp/sdist-with.txt  # nothing
$ diff /tmp/wheel-without.txt /tmp/wheel-with.txt  # nothing
```

Reply:
> i.e. the final path component (`bar`) is the name of the package as included in the final wheel (because it's taken relative to `sources`).

So is the reason that removing

```toml
[tool.hatch.build.targets.wheel]
packages = ["src/pyhf"]
```

didn't have any effect is because we're using `only-include` for `[tool.hatch.build.targets.sdist]` and we only have one package directory in `src/`?

Reply:
OK, the original issue that we were fixing with `only-include` has been fixed (I think) by copying the schemas into the docs at docs time: https://github.com/scikit-hep/pyhf/commit/81a429ec8888263d69a97da40ff03da5ed4f6dda

So we don't need this explicitly. I found that reverting `9ca5b12` didn't leave the schemas out of the sdist or wheel for me. If you revert `9ca5b12` and keep `4a4da3e`, what do you see?

Reply:
Configuration of build targets doesn't conflict with nor influence other targets. The reason omitting that from the wheel results in the same contents is because the file `src/pyhf/__init__.py` exists and therefore Hatchling determines that the default inclusion option for the wheel target should be as you had it `packages = ["src/pyhf"]`

- https://hatch.pypa.io/latest/plugins/builder/wheel/#default-file-selection
- https://github.com/pypa/hatch/blob/hatchling-v1.12.2/backend/src/hatchling/builders/wheel.py#L172-L200

Reply:
Yep, I think we're all on the same page! @ofek serious question, do you ever sleep? :eyes: ;)

Reply:
> OK, the original issue that we were fixing with `only-include` has been fixed (I think) by copying the schemas into the docs at docs time

No, copying the schemas into the docs was added way back in PR #458.

> i.e., I think the default Hatch build configuration should pick everything up. But, if you want to use e.g. only-include, you can for other reasons

Unfortunately no. :/ If we look at the changes between now (`f3ebf0e6`) and before we started poking at `[tool.hatch.build.targets.wheel]` when @kratsg wondered if it was needed

```diff
$ git diff 1bf9c0e9 f3ebf0e6 
diff --git a/pyproject.toml b/pyproject.toml
index 90e2d9c7..73aebb4c 100644
--- a/pyproject.toml
+++ b/pyproject.toml
@@ -130,7 +130,9 @@ version-file = "src/pyhf/_version.py"
 exclude = [
     "/src/conftest.py"
 ]
-include = [
+# only-include needed to properly include src/pyhf/schemas
+# c.f. https://github.com/pypa/hatch/pull/299
+only-include = [
     "/src",
     "/LICENSE",
     "/README.rst",
@@ -139,9 +141,6 @@ include = [
     "/CITATION.cff"
 ]
 
-[tool.hatch.build.targets.wheel]
-packages = ["src/pyhf"]
-
 [tool.black]
 line-length = 88
 target-version = ['py38', 'py39', 'py310']
```

```
$ git reset --hard 1bf9c0e9
```

then we have

```toml
...

[tool.hatch.build.targets.sdist]
exclude = [
    "/src/conftest.py"
]
include = [
    "/src",
    "/LICENSE",
    "/README.rst",
    "/pyproject.toml",
    "/AUTHORS",
    "/CITATION.cff"
]

[tool.hatch.build.targets.wheel]
packages = ["src/pyhf"]

...
```

With this we are not getting the `schemas/` directory included in the sdist

```console
$ rm -rf dist && python -m build . && python -m tarfile --list dist/pyhf*.tar.gz | grep schemas  # nothing
```

If we now 

```
git cherry-pick 9ca5b123
```

to use `only-include` then the schemas are included

```console
$ rm -rf dist && python -m build . && python -m tarfile --list dist/pyhf*.tar.gz | grep schemas
...
pyhf-0.7.1.dev66/src/pyhf/schemas/HistFactorySchema.dtd 
pyhf-0.7.1.dev66/src/pyhf/schemas/1.0.0/defs.json 
pyhf-0.7.1.dev66/src/pyhf/schemas/1.0.0/jsonpatch.json 
pyhf-0.7.1.dev66/src/pyhf/schemas/1.0.0/measurement.json 
pyhf-0.7.1.dev66/src/pyhf/schemas/1.0.0/model.json 
pyhf-0.7.1.dev66/src/pyhf/schemas/1.0.0/patchset.json 
pyhf-0.7.1.dev66/src/pyhf/schemas/1.0.0/workspace.json
```

regardless of if we have 

```toml
[tool.hatch.build.targets.wheel]
packages = ["src/pyhf"]
```

or not

```console
$ git reset --hard f3ebf0e6
$ rm -rf dist && python -m build . && python -m tarfile --list dist/pyhf*.tar.gz | grep schemas
...
pyhf-0.7.1.dev68/src/pyhf/schemas/HistFactorySchema.dtd 
pyhf-0.7.1.dev68/src/pyhf/schemas/1.0.0/defs.json 
pyhf-0.7.1.dev68/src/pyhf/schemas/1.0.0/jsonpatch.json 
pyhf-0.7.1.dev68/src/pyhf/schemas/1.0.0/measurement.json 
pyhf-0.7.1.dev68/src/pyhf/schemas/1.0.0/model.json 
pyhf-0.7.1.dev68/src/pyhf/schemas/1.0.0/patchset.json 
pyhf-0.7.1.dev68/src/pyhf/schemas/1.0.0/workspace.json
```

So I'm a bit confused as to why it doesn't have any effect even when building without an sdist first

```
# This is without it
$ rm -rf dist && python -m build --wheel . && python -m zipfile --list dist/pyhf*.whl | grep schemas
* Creating virtualenv isolated environment...
* Installing packages in isolated environment... (hatch-vcs, hatchling>=1.0.0)
* Getting build dependencies for wheel...
* Building wheel...
Successfully built pyhf-0.7.1.dev68-py3-none-any.whl
pyhf/schemas/HistFactorySchema.dtd             2020-02-02 00:00:00         6521
pyhf/schemas/1.0.0/defs.json                   2020-02-02 00:00:00        13069
pyhf/schemas/1.0.0/jsonpatch.json              2020-02-02 00:00:00          180
pyhf/schemas/1.0.0/measurement.json            2020-02-02 00:00:00          184
pyhf/schemas/1.0.0/model.json                  2020-02-02 00:00:00          172
pyhf/schemas/1.0.0/patchset.json               2020-02-02 00:00:00          178
pyhf/schemas/1.0.0/workspace.json              2020-02-02 00:00:00          180
```

but if you would suggest putting it in just to guard in the future I'm okay with doing so if @kratsg doesn't have strong objections.

Reply:
...thought it might be worth noting that @henryiii's `cookie` doesn't use `tool.hatch.build` at all

```console
...
(venv) root@882d0e43004d:/# cookiecutter gh:scikit-hep/cookie
project_name [package]: 
org [Scikit-HEP]: 
url [https://github.com/Scikit-HEP/package]: 
full_name [My Name]: 
Select project_type:
1 - hatch
2 - setuptools
3 - setuptools621
4 - pybind11
5 - skbuild
6 - poetry
7 - flit
8 - pdm
9 - trampolim
10 - whey
11 - maturin
Choose from 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11 [1]: 1
email [me@email.com]: 
project_short_description [A great package.]: 
year [2023]: 
Select license:
1 - BSD
2 - Apache
3 - MIT
Choose from 1, 2, 3 [1]: 
(venv) root@882d0e43004d:/# grep -r "tool.hatch" package/
package/pyproject.toml:[tool.hatch]
...
```

Reply:
You're right, I didn't interrogate the blame deeply enough. Probably just the move to nox registered in my mind as "added in this commit". OK.

I did some more digging, and indeed we still have the same problem as the original PR: `docs/_extras/schemas` is a symlink, so it would get resolved, and then the `pyhf/schemas` would be missing. Indeed, I only see the docs schemas when I delete `only-include`.

So, we need either `packages` or `only-include` here, unless we decide to delete the symlink in the repo. This would mean that the sdist would not include these schemas, though, which might not be desirable. 

@matthewfeickert can we delete these and have the `docs` target copy them too?

Reply:
> If you don't pass any arguments it builds the source distribution then unpacks that to build the wheel whereas if you pass the flag to just build the wheel it will do so from the project root directly

Yes, this was clear to me. Sorry, I should have mentioned something like the last bit of my comment in https://github.com/scikit-hep/pyhf/pull/2095#discussion_r1082010630 earlier or made this more explicit than just grepping the sdist contents. :grimacing: 

> Configuration of build targets doesn't conflict with nor influence other targets. The reason omitting that from the wheel results in the same contents is because the file `src/pyhf/__init__.py` exists and therefore Hatchling determines that the default inclusion option for the wheel target should be as you had it `packages = ["src/pyhf"]`
> 
> * https://hatch.pypa.io/latest/plugins/builder/wheel/#default-file-selection
> * https://github.com/pypa/hatch/blob/hatchling-v1.12.2/backend/src/hatchling/builders/wheel.py#L172-L200

..and _this_ answers my confused questions about `[tool.hatch.build.targets.wheel]`. :+1: Sorry I should have read the docs much more carefully.

> When the **user has not set any [file selection](https://hatch.pypa.io/latest/config/build/#file-selection) options**, the [project name](https://hatch.pypa.io/latest/config/metadata/#name) will be used to determine the package to ship in the following heuristic order

...and this also motivates me to say that even though it isn't _necessary_ it is still worth including

```toml
[tool.hatch.build.targets.wheel]
packages = ["src/pyhf"]
```

as it means that `hatchling` doesn't have to do any additional logic checks to figure anything out. :tada: 

So I'm going to add that back in. :+1: 

Thanks very much for your patience @agoose77 and thank you very much for your time and help @ofek! :pray:

Reply:
@matthewfeickert I think you should still need that option. Hatch, for example, has `hatch build`, and PEP 517 says
> To ensure that wheels from different sources are built the same way, frontends may call build_sdist first, and then call build_wheel in the unpacked sdist. But if the backend indicates that it is missing some requirements for creating an sdist (see below), the frontend will fall back to calling build_wheel in the source directory.

I presume @ofek that `hatch build` is directly generating a wheel without an intermediate sdist, right? (If I drop the `packages` configuration for `targets.wheel`, the missing-schemas phenomenon (as a symptom for no `include-only`/`packages` ) appears again). Perhaps we should change that in Hatch, as Hatch is a PEP 517 frontend (technically, even if only `hatchling` is supported right now)? If you agree, I can open an issue and get back to it in a couple of months.

Regardless of whether hatch will render this unneeded, let's keep it.

Reply:
> It doesn't proscribe that one _must_ build a wheel from an sdist.

Correct. At the end of my rambling comment https://github.com/scikit-hep/pyhf/pull/2095#discussion_r1082010630 I showed (at the "So I'm a bit confused as to why it doesn't have any effect even when building without an sdist first" part) that without building an sdist first 

```console
$ python -m build --wheel .
```

would correctly pickup the `schemas/` dir regardless of if 

```toml
[tool.hatch.build.targets.wheel]
packages = ["src/pyhf"]
```

was present or not thanks to hatchling's default file selection working as intended (https://github.com/scikit-hep/pyhf/pull/2095#discussion_r1081998703).

> Regardless of whether hatch will render this unneeded, let's keep it.

Agreed.

Reply:
I'm just being verbose here for ease. This might be your understanding too, but I want to be sure we are all on the same page! :)

Python `build --wheel` is doing the recommended thing of `sdist → wheel`. Hatch's default file inclusion logic always picks up the `pyhf/schemas` directory, but the symlinked schemas in `docs/_extras/` cause this to fail. Eliminating these from the SDist (manually, with `only-include` et al.) means that the wheel builder then succeeds with default inclusion logic, assuming that indeed we go `sdist → wheel`. 

Meanwhile, `hatch build --target wheel` doesn't do `sdist → wheel`, and therefore the wheel build suffers the same problem as the sdist; the schemas are found, but encountered under the wrong (excluded) path (`docs/_extras`)

So we definitely need an `only-include` like directive in both `targets.sdist` and `targets.wheel` whilst we have these duplicate schemas.

Reply:
> Python `build --wheel` is doing the recommended thing of `sdist → wheel`.

I don't think that's exactly correct. From `python -m build --help`:

```
    By default, a source distribution (sdist) is built from {srcdir}
    and a binary distribution (wheel) is built from the sdist.
    This is recommended as it will ensure the sdist can be used
    to build wheels.

    Pass -s/--sdist and/or -w/--wheel to build a specific distribution.
    If you do this, the default behavior will be disabled, and all
    artifacts will be built from {srcdir} (even if you combine
    -w/--wheel with -s/--sdist, the wheel will be built from {srcdir}).
```

so my `python -m build --wheel .` is building from the srcdir, and not from sdist (the default behavior).

Reply:
Gah, then "`Python build --wheel` is _not_ doing the recommended thing of `sdist → wheel`" :laughing: 

Admission time ... I discovered my environment didn't fully upgrade; I was evaluating `hatch` behaviour in-part based upon CLI exploration. I upgraded `hatchling` to the newest version explicitly, and the different results that I was seeing evaporated. I had been confused that we seemed to be drawing different conclusions.

Now Ofek's points make more sense — for the SDist, we need to curtail the default search for the above reasons, but the wheel builder works fine due to the comment from Ofek that you linked!

So my view is different now; we don't need that setting, and we only added it because it fixes an annoying build problem. As long as hatchling's PEP 517 hook produces a good wheel when used without an intermediate sdist (which it does), it should be fine to remove it. So, your call, but I think you can safely remove it!

Sorry for the wild goose chase. I had thought I was being helpful, but with the above context I can see it was actually just confusing things!

Reply:
Nothing to apologize about, you have been beyond helpful. :)

> Now Ofek's points make more sense — for the SDist, we need to curtail the default search for the above reasons, but the wheel builder works fine due to the comment from Ofek that you linked!

:heavy_check_mark: :+1: We are in agreement on understanding now!

> As long as hatchling's PEP 517 hook produces a good wheel when used without an intermediate sdist (which it does), it should be fine to remove it. So, your call, but I think you can safely remove it!

We _can_ remove it but keeping it in skips all of

https://github.com/pypa/hatch/blob/hatchling-v1.12.2/backend/src/hatchling/builders/wheel.py#L172-L200

which while I'm sure is only a few microseconds, seems like a "why not just leave in 2 lines?" question to me now.

Reply:
Hatchling currently does not build the wheel from the source distribution. I plan to do that but I don't have time currently. It would actually be useful if you did open an issue so I can track it